src/downloads_counter.rs

```diff
@@ -0,0 +1,154 @@
+use crate::App;
+use anyhow::Error;
+use dashmap::{DashMap, SharedValue};
+use diesel::{pg::upsert::excluded, prelude::*};
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicI64, AtomicUsize, Ordering};
+
+/// crates.io receives a lot of download requests, and we can't execute a write query to the
+/// database during each connection for performance reasons. To reduce the write load, this struct
+/// collects the pending updates from the current process and writes in batch.
+///
+/// To avoid locking the whole data structure behind a RwLock, which could potentially delay
+/// requests, this uses the dashmap crate. A DashMap has the same public API as an HashMap, but
+/// stores the items into `num_cpus()*4` individually locked shards. This approach reduces the
+/// likelyhood of a request encountering a locked shard.
+///
+/// Persisting the download counts in the database also takes advantage of the inner sharding of
+/// DashMaps: to avoid locking all the download requests at the same time each iteration only
+/// persists a single shard at the time.
+///
+/// The disadvantage of this approach is that download counts are stored in memory until they're
+/// persisted, so it's possible to lose some of them if the process exits ungracefully. While
+/// that's far from ideal, the advantage of batching database updates far outweights potentially
+/// losing some download counts.
+#[derive(Debug)]
+pub struct DownloadsCounter {
+    /// Inner storage for the download counts.
+    inner: DashMap<i32, AtomicUsize>,
+    /// Index of the next shard that should be persisted by `persist_next_shard`.
+    shard_idx: AtomicUsize,
+    /// Number of downloads that are not yet persisted on the database. This is just used as a
+    /// metric included in log lines, and it's not guaranteed to be accurate.
+    pending_count: AtomicI64,
+}
+
+impl DownloadsCounter {
+    pub(crate) fn new() -> Self {
+        Self {
+            inner: DashMap::new(),
+            shard_idx: AtomicUsize::new(0),
+            pending_count: AtomicI64::new(0),
+        }
+    }
+
+    pub(crate) fn increment(&self, version_id: i32) {
+        self.pending_count.fetch_add(1, Ordering::SeqCst);
+
+        if let Some(counter) = self.inner.get(&version_id) {
+            counter.value().fetch_add(1, Ordering::SeqCst);
+        } else {
+            self.inner
+                .entry(version_id)
+                .and_modify(|counter| {
+                    counter.fetch_add(1, Ordering::SeqCst);
```

Comment:
So if I understand correctly, this is very similar to the `if let ... inner.get()` above. However, `get()` returns an immutable reference, while `and_modify()` passes a `&mut` to the closure (which is more strict than we really need).

The "redundant" `fetch_add()` is needed here, in case we race with another thread which inserts the version between our calls to `get()` and `entry()`. And we use the `get()` branch above for the happy path where we don't need to lock the shard.

Reply:
Yes, your understanding is correct! I'll add comments to write it down in the source code.

Reply:
Added some comments on the `increment` function.